### DISCUSSION TITLE
SD Card problem while installation

### DISCUSSION THREAD

**Discussion by Rak3tenmann** [jomjol/AI-on-the-edge-device#2710] [General]
Hi there! I tried 2 different SD-Cards during the setup process. A 32 GB and a 2 GB. Same issue.
WIth both, I can setup the initial copy via the browser interface. Which makes me beliefe, the cards are writebale with the ESP. Whole copy process works fine. However, after rebooting it comes up with 4 blink which indicates a SD Card problem, if I am not mistaking. 
WOuld it realy make sense do buy another one? Or is there another possible problem.
(Format is FAT / MBR)

**Comment by caco3** [comment]
You could try to format the SD cards differently.
Search the discussions for it, there where several MAC users who reported similar issues.

**Comment by SybexX** [comment]
Since version 15.2.0, ESP-IDF 5.0.1 has been used, where the value "FF_USE_TRIM" was set from 0 to 1 in the framework-espidf\components\fatfs\src\ffconf.h and this causes a lot of problems with SD detection and read/write access.
After I set the value back to 0 in all versions (ESP-IDF > 4.4.5), there are no longer any problems with the SD card, no matter which one I use.

**Reply by caco3** [reply]
Do you have a link to the documentation of this flag?
Can you show us your change or provide a Pull Request?

I can see it is used here: https://github.com/espressif/esp-idf/blob/master/components/fatfs/diskio/diskio_sdmmc.c#L74, but I can not find any documentation for it using google or at https://docs.espressif.com/projects/esp-idf/en/latest/esp32/search.html?q=FF_USE_TRIM

**Reply by SybexX** [reply]
I changed the value to 0 in the file C:\Users\Michael\ .platformio\packages\framework-espidf@3.50002.230601\components\fatfs\src\ffconf.h manuel.
This framework is automatically downloaded when compiling with VS-Code.

#define FF_USE_TRIM		1
/* This option switches support for ATA-TRIM. (0:Disable or 1:Enable)

If you search for "ATA-TRIM" on Google, you'll find a lot about it.

**Reply by SybexX** [reply]
https://github.com/espressif/esp-idf/issues/8704

**Reply by caco3** [reply]
Your change is not part of your code but part of the ESP IDF Framework.
See https://github.com/espressif/esp-idf/blob/master/components/fatfs/src/ffconf.h#L236
We can not simply apply those changes. We need to find another way to fix this...

**Reply by SybexX** [reply]
That's the quickest and easiest way^^

An alternative would be to copy the "fatfs" folder from the ESP IDF Framework into the "components" folder (of course with the change #define FF_USE_TRIM 0) and adjust the #include in the main.cpp.

#include "esp_vfs_fat.h" -----> #include "../../components/fatfs/vfs/esp_vfs_fat.h"
#include "ffconf.h" -----> #include "../../components/fatfs/src/ffconf.h"

I know it doesn't really solve the problem, but this is the quickest way to get it working until you find a proper solution.

**Reply by caco3** [reply]
Your other approach did not work I guess (the changes you did to main.c)?
Before I go that way i would like to have a feedback from @friedpa if it also works for him.

**Reply by SybexX** [reply]
![aenderung1](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/acd229da-9c15-4fc5-9989-36fe5633ff17)
![aenderung2](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/2013dc09-9efd-49cf-ab7c-4ffa2438e03b)

I did it like that now and it works.
It's no different with "miniz" from the "jomjol_fileserver_ota" folder, which looks like it was also copied from the ESP IDF Framework.
I tested the whole thing with three noname SDs (2x 128MB and 1x 4GB) and everything works as it should^^

**Comment by friedpa** [comment]
This is very interesting as a lot of problems have been reported in this area since 15.2.0
@jomjol @caco3 could you please do a new rolling version with this changed parameter to try out the finding. I would then use my former non-compatible SD cards in my test system to try it out and report back accordingly.

**Comment by SybexX** [comment]
[AI-on-the-edge-device__update__rolling.zip](https://github.com/jomjol/AI-on-the-edge-device/files/13777174/AI-on-the-edge-device__update__rolling.zip)

**Reply by friedpa** [reply]
Did you use 15.4.0 as basic SW version?

**Reply by SybexX** [reply]
[AI-on-the-edge-device__update__rolling.zip](https://github.com/jomjol/AI-on-the-edge-device/files/13779723/AI-on-the-edge-device__update__rolling.zip)
this time compiled with version.
The first time I just downloaded it and didn't clone it, that's why it was NaN^^

**Comment by SybexX** [comment]
I used the current AI-on-the-edge-device-rolling

**Comment by friedpa** [comment]
So I did this morning the tests and surprisingly from my 5 not working SDs 3 did work now. For the remaining two, one is also not readable in my Laptop, so one is left, which is working in my Laptop but not in the ESP.

**Comment by caco3** [comment]
Thanks @SybexX for the proposal and also @friedpa for the validation.
I tried to find a better way than copying the component, but that seems to be the best aproach for now.
See https://github.com/jomjol/AI-on-the-edge-device/pull/2779 for the fix and https://github.com/jomjol/AI-on-the-edge-device/actions?query=branch%3AFF_USE_TRIM+event%3Apull_request for a build.
The log will show the `FF_USE_TRIM` value and a warning if it is set to `1`.

**Reply by SybexX** [reply]
I found the error in the ESP-IDF:

With "#define FF_USE_TRIM 1" all ATA-Trim routines are activated and run whether the SD/MMC supports it or not.
If the SD/MMC does not support ATA-Trim, a write error occurs and the SD/MMC initialization is aborted.
There is a function to check ATA-Trim support, but it is useless.

I have adapted the files responsible for this so that the SDs/MMCs with ATA trim support can also use it.
However, couldn't check it as I don't have any with ATA-Trim support here.
[components.zip](https://github.com/jomjol/AI-on-the-edge-device/files/13797421/components.zip)

I tried your FF_USE_TRIM Branch, but without success because my PlatformIO still uses the fatfs from the ESP-IDF and not from the components folder.

**Reply by caco3** [reply]
Could you create a pull request? that would make it easier

> I tried your FF_USE_TRIM Branch, but without success because my PlatformIO still uses the fatfs from the ESP-IDF and not from the components folder.

How do you know it does not work? Is FF_USE_TRIM still set to 1?
You could also use the prebuild image from the 2nd link.

**Reply by SybexX** [reply]
In the ESP-IDF I had FF_USE_TRIM set to 1 and in the components folder set to 0. After compiling and installing, the ESP did not recognize my SD.
If I set FF_USE_TRIM to 0 in both, it works.
Who knows, maybe an extension I installed in VScode is causing the linking problems^^

**Comment by caco3** [comment]
https://github.com/jomjol/AI-on-the-edge-device/pull/2781 brings improvement to SD card support, thanks to @SybexX
